extensions/blocks/publicize/store/selectors.js

```diff
@@ -1,24 +1,257 @@
+ */
+export function getTweetTemplate( state ) {
+	const twitterAccount = state.connections?.find(
+		connection => 'twitter' === connection.service_name
+	);
+
+	return {
+		date: Date.now(),
+		name: twitterAccount?.profile_display_name || __( 'Account Name', 'jetpack' ),
+		profileImage:
+			twitterAccount?.profile_picture ||
+			'https://abs.twimg.com/sticky/default_profile_images/default_profile_bigger.png',
+		screenName: twitterAccount?.display_name || '',
+	};
+}
+
+/**
+ * Generates an array of tweets, including Twitter account data.
+ *
+ * @param {object} state - State object.
+ *
+ * @returns {Array} Array of tweets.
+ */
+export function getTweetStorm( state ) {
+	const tweetTemplate = getTweetTemplate( state );
+
+	const thread = [
+		getFirstTweet( state ),
+		...state.tweets.map( tweet => ( {
+			...tweetTemplate,
+			text: tweet.text,
+			media: tweet.media,
+			tweet: tweet.tweet,
+			urls: tweet.urls,
+			card: getTwitterCardForURLs( state, tweet.urls ),
+		} ) ),
+	];
+
+	// Only add the last tweet if there's actual content in the thread.
+	if ( thread.length > 1 ) {
+		thread.push( getLastTweet( state ) );
+	}
+
+	return thread;
+}
+
+/**
+ * Constructs the first tweet to use in the thread.
+ *
+ * @param {object} state - State object.
+ *
+ * @returns {object} The tweet.
+ */
+export function getFirstTweet( state ) {
+	// This isn't defined properly in the test environment, so we have to skip this function.
+	if ( ! select( 'core' ) ) {
+		return;
+	}
+
+	const tweetTemplate = getTweetTemplate( state );
+
+	const { getMedia } = select( 'core' );
+	const { getEditedPostAttribute } = select( 'core/editor' );
+
+	const featuredImageId = getEditedPostAttribute( 'featured_media' );
+	const url = getEditedPostAttribute( 'link' );
+
+	const media = featuredImageId && getMedia( featuredImageId );
+	const image = media?.media_details?.sizes?.large?.source_url || media?.source_url;
+
+	return {
+		...tweetTemplate,
+		text: getShareMessage() + ` ${ url }`,
+		urls: [ url ],
+		card: {
+			title: getEditedPostAttribute( 'title' ),
+			description:
+				getEditedPostAttribute( 'meta' )?.advanced_seo_description ||
+				getEditedPostAttribute( 'excerpt' ) ||
+				getEditedPostAttribute( 'content' ).split( '<!--more' )[ 0 ] ||
+				__( 'Visit the post for more.', 'jetpack' ),
+			url,
+			image,
+			type: image ? 'summary_large_image' : 'summary',
+		},
+	};
+}
+
+/**
+ * Constructs the last tweet to use in the thread.
+ *
+ * @param {object} state - State object.
+ *
+ * @returns {object} The tweet.
+ */
+export function getLastTweet( state ) {
+	// This isn't defined properly in the test environment, so we have to skip this function.
+	if ( ! select( 'core/editor' ) ) {
+		return;
+	}
+
+	const { getEditedPostAttribute } = select( 'core/editor' );
+	const url = getEditedPostAttribute( 'link' );
+
+	return {
+		...getFirstTweet( state ),
+		text: __( "I've also published this thread on my site:", 'jetpack' ) + ` ${ url }`,
```

Comment:
Ah, I should've documented that. 🙂 The URL _must_ be the last thing in the tweet, so that Twitter will trim it from the tweet text when it displays that tweet.

I've added a comment in b001bf38d to note that.